src/coreclr/nativeaot/Runtime/riscv64/WriteBarriers.S

```diff
@@ -220,31 +210,31 @@ LEAF_END RhpByRefAssignRef, _TEXT
 // on the managed heap.
 //
 // On entry:
-//   t5 : the destination address (LHS of the assignment).
+//   t3 : the destination address (LHS of the assignment).
 //         May not be a heap location (hence the checked).
-//   t6 : the object reference (RHS of the assignment).
+//   t4 : the object reference (RHS of the assignment).
 //
 // On exit:
-//   t3, t4 : trashed
-//   t5      : incremented by 8
+//   t2, t6 : trashed
+//   t3      : incremented by 8
 LEAF_ENTRY RhpCheckedAssignRef, _TEXT
 
-        // is destReg within the heap?
-        la    t3, g_lowest_address
-        ld    t3, 0(t3)
-        sltu  t4, t6, t3
+        # Check if the destination is within the heap bounds
+        la t2, C_FUNC(g_lowest_address)
+        la t6, C_FUNC(g_highest_address)
 
-        la    t3, g_highest_address
-        ld    t3, 0(t3)
-        sltu  t0, t3, t6
-        or    t4, t0, t4
-        beq   t4, zero, C_FUNC(RhpAssignRefRiscV64)
+        bltu    t3, t2, LOCAL_LABEL(NotInHeap)
+        bgeu    t3, t6, LOCAL_LABEL(NotInHeap)
 
-NotInHeap:
-    ALTERNATE_ENTRY RhpCheckedAssignRefAVLocation
-        sd    t6, 0(t5)
-        addi  t5, t5, 8
-        jalr  ra
+        j       C_FUNC(RhpAssignRefRiscV64)
+
+LOCAL_LABEL(NotInHeap):
+        ALTERNATE_ENTRY RhpCheckedAssignRefAVLocation
+        sd      t4, 0(t3)
+        addi    t3, t3, 8
+
+        // Jump to the return address in a1 for the tailcall.
```

Comment:
```suggestion
        // TODO-RISCV64: Jump to the return address in a1 for the tailcall. This should be revert to ret instead of jr once JIT is fixed
```